micrometer-tracing/src/main/java/io/micrometer/tracing/contextpropagation/ObservationAwareSpanThreadLocalAccessor.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2023 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.tracing.contextpropagation;
+
+import io.micrometer.context.ThreadLocalAccessor;
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.ObservationRegistry;
+import io.micrometer.observation.contextpropagation.ObservationThreadLocalAccessor;
+import io.micrometer.tracing.Span;
+import io.micrometer.tracing.Tracer;
+import io.micrometer.tracing.handler.TracingObservationHandler;
+
+/**
+ * A {@link ThreadLocalAccessor} to put and restore current {@link Span} depending on
+ * whether {@link ObservationThreadLocalAccessor} did some work or not (if
+ * {@link ObservationThreadLocalAccessor} opened a scope, then this class doesn't want to
+ * create yet another span).
+ *
+ * In essence logic of this class is as follows:
+ *
+ * <ul>
+ * <li>If {@link ObservationThreadLocalAccessor} created a {@link Span} via the
+ * {@link TracingObservationHandler} - do nothing</li>
+ * <li>else - take care of the creation of a {@link Span} and putting it in thread
+ * local</li>
+ * </ul>
+ *
```

Comment:
nit: better to mention the ordering that this accessor needs to be registered after `ObservationThreadLocalAccessor`.